### ISSUE TITLE
SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER  mode issue

### ISSUE THREAD

**Issue by QinLongFei** [openssl/openssl#5299]
If application behave incorrectly with SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER  mode, openssl will crash and more than 16K unexpected data will be sent to peer.

For example, there are two buffers need to send to peer. buf1 is 1024 bytes and buf2 is 64 bytes.

...

SSL_set_mode(s,SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER);

SSL_write(s, buf1, 1024);

For some reason, only 512 bytes write successfully and application not check the return value but continue to write buf2.

SSL_write(s, buf2, 64);

Based on ssl3_write_bytes() logic, due to we set SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER mode,  the left 512 bytes will write successfully(even though the buffer address has changed from buf1 to buf2) and then contine to write buf2. In this point, n will be explained as a big unsigned number. It will lead to openssl send much data to peer still some crash happen.

It's dangerous for someone who behave SSL_write() incorrectly. We should fix this issue or at least, give some alert warning.

**Comment by bernd-edlinger** [comment]
You need to check the return code of SSL_write.
When the write has not completed, you may move buf1 to somewhere else,
but you may not change the data, or make it shorter as in your example.

**Comment by mattcaswell** [comment]
Right. I think the point is though that we could help prevent the user from using this particular foot gun by validating that the length has remained constant.

**Comment by bernd-edlinger** [comment]
Hmm, I wonder how n can become negative in ssl3_write_bytes after this check:
```
    if ((len < s->rlayer.wnum)
        || ((wb->left != 0) && (len < (s->rlayer.wnum + s->rlayer.wpend_tot)))) {
        SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_SSL3_WRITE_BYTES,
                 SSL_R_BAD_LENGTH);
        return -1;
    }
```

**Comment by mattcaswell** [comment]
I've just been looking at the same piece of code trying to figure out how this could happen!

**Comment by mattcaswell** [comment]
@QinLongFei what version of OpenSSL did you observe this behaviour on?

**Comment by bernd-edlinger** [comment]
One minor nit I see in ssl3_write_pending:

```
        || ((s->rlayer.wpend_buf != buf) &&
            !(s->mode & SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER))
```

Suppose wpend_buf is the old buf and it has been free'd since the last call to SSL_write
then this pointer will become indeterminate, therefore this should not be used at all in
SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER.

**Comment by mattcaswell** [comment]
So that clause should be written the other way around (with the `s>mode` check first)

**Comment by bernd-edlinger** [comment]
Yes, I will send a PR for this.

**Comment by QinLongFei** [comment]
I found and verified it on OpenSSL-1.0.2k.

Thanks

Get Outlook for iOS<https://aka.ms/o0ukef>

**Comment by QinLongFei** [comment]
Code  in ssl3_write_pending(),  

((s->s3->wpend_buf != buf) &&
            !(s->mode & SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER))

if SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER is set, openssl will ignore (s->s3->wpend_buf != buf) , the paritial wirte will alway do even though buffer has changed. 

Code in  ssl3_write_bytes(),

n = (len - tot);
    for (;;) {
        if (n > s->max_send_fragment)
            nw = s->max_send_fragment;
        else
            nw = n;

        i = do_ssl3_write(s, type, &(buf[tot]), nw, 0);
        if (i <= 0) {
            /* XXX should we ssl3_release_write_buffer if i<0? */
            s->s3->wnum = tot;
            return i;
        }

if len<tot, n will be  explained  as a big unsigned number.

As my example said, after completing partial write for buf1, tot=1024, but len=64.

I think we had better do some check between len and tot.

**Comment by mattcaswell** [comment]
Ah. 1.0.2k is a little old. Looks like this was already fixed from 1.0.2l onwards by commit 71d66c46c7. See also #3122.
